File: week-04-fastapi/app/routers/categories.py
from fastapi import APIRouter, HTTPException
from bson import ObjectId
from app.database import get_db
from app.schemas import CategoryCreate, CategoryUpdate, CategoryResponse

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get("", response_model=list[CategoryResponse])
def list_categories():
    db = get_db()
    cats = db.categories.find().sort("name", 1)
    return [CategoryResponse(**c) for c in cats]


@router.get("/{cat_id}", response_model=CategoryResponse)
def get_category(cat_id: str):
    if not ObjectId.is_valid(cat_id):
        raise HTTPException(400, "Invalid category ID")
    db = get_db()
    cat = db.categories.find_one({"_id": ObjectId(cat_id)})
    if not cat:
        raise HTTPException(404, "Category not found")
    return CategoryResponse(**cat)


@router.post("", response_model=CategoryResponse, status_code=201)
def create_category(body: CategoryCreate):
    db = get_db()
    result = db.categories.insert_one(body.model_dump())
    cat = db.categories.find_one({"_id": result.inserted_id})
    return CategoryResponse(**cat)


@router.put("/{cat_id}", response_model=CategoryResponse)
def update_category(cat_id: str, body: CategoryUpdate):
    if not ObjectId.is_valid(cat_id):
        raise HTTPException(400, "Invalid category ID")
    db = get_db()
    result = db.categories.update_one(
        {"_id": ObjectId(cat_id)},
        {"$set": body.model_dump()}
    )
    if result.matched_count == 0:
        raise HTTPException(404, "Category not found")
    cat = db.categories.find_one({"_id": ObjectId(cat_id)})
    return CategoryResponse(**cat)


@router.delete("/{cat_id}", status_code=204)
def delete_category(cat_id: str):
    if not ObjectId.is_valid(cat_id):
        raise HTTPException(400, "Invalid category ID")
    db = get_db()
    result = db.categories.delete_one({"_id": ObjectId(cat_id)})
    if result.deleted_count == 0:
        raise HTTPException(404, "Category not found")

File: week-05-express/src/routes/categories.js
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../db.js";

const router = Router();

router.get("/", async (req, res) => {
  const db = getDb();
  const cats = await db.collection("categories").find().sort({ name: 1 }).toArray();
  res.json(cats.map(c => ({ id: c._id.toString(), name: c.name, slug: c.slug })));
});

router.get("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid category ID" });
  const db = getDb();
  const cat = await db.collection("categories").findOne({ _id: new ObjectId(req.params.id) });
  if (!cat) return res.status(404).json({ detail: "Category not found" });
  res.json({ id: cat._id.toString(), name: cat.name, slug: cat.slug });
});

router.post("/", async (req, res) => {
  const db = getDb();
  const result = await db.collection("categories").insertOne({ name: req.body.name, slug: req.body.slug });
  const cat = await db.collection("categories").findOne({ _id: result.insertedId });
  res.status(201).json({ id: cat._id.toString(), name: cat.name, slug: cat.slug });
});

router.put("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid category ID" });
  const db = getDb();
  const result = await db.collection("categories").updateOne(
    { _id: new ObjectId(req.params.id) },
    { $set: { name: req.body.name, slug: req.body.slug } }
  );
  if (result.matchedCount === 0) return res.status(404).json({ detail: "Category not found" });
  const cat = await db.collection("categories").findOne({ _id: new ObjectId(req.params.id) });
  res.json({ id: cat._id.toString(), name: cat.name, slug: cat.slug });
});

router.delete("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid category ID" });
  const db = getDb();
  const result = await db.collection("categories").deleteOne({ _id: new ObjectId(req.params.id) });
  if (result.deletedCount === 0) return res.status(404).json({ detail: "Category not found" });
  res.status(204).send();
});

export default router;
